extensions/amp-story-360/0.1/amp-story-360.js

```diff
@@ -0,0 +1,291 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CSS} from '../../../build/amp-story-360-0.1.css';
+import {timeStrToMillis} from '../../../extensions/amp-story/1.0/utils';
+import {CommonSignals} from '../../../src/common-signals';
+import {whenUpgradedToCustomElement} from '../../../src/dom';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {userAssert} from '../../../src/log';
+import {Services} from '../../../src/services';
+import {Matrix, Renderer} from '../../../third_party/zuho/zuho';
+
+
+/**
+ * Internal helper class representing a camera orientation (POV) in polar
+ * coordinates.
+ */
+class CameraOrientation {
+  /**
+   * @param {number} theta
+   * @param {number} phi
+   * @param {number} scale
+   */
+  constructor(theta, phi, scale) {
+    this.theta = theta;
+    this.phi = phi;
+    this.scale = scale;
+  }
+
+  /**
+   * @param {number} heading
+   * @param {number} pitch
+   * @param {number} zoom
+   * @return {!CameraOrientation}
+   */
+  static fromDegrees(heading, pitch, zoom) {
+    const deg2rad = (deg) => deg * Math.PI / 180;
+    return new CameraOrientation(
+        deg2rad(-pitch - 90), deg2rad(heading), 1 / zoom);
+  }
+
+  /**
+   * @return {!Float32Array}
+   */
+  get rotation() {
+    return Matrix.mul(
+        3, Matrix.rotation(3, 1, 2, this.theta),
+        Matrix.rotation(3, 0, 1, this.phi));
+  }
+}
+
+
+/**
+ * Internal helper class generating a sequence of frame by frame orientations
+ * from a set of 2 or more CameraOrientation target points and an animation
+ * duration (approximate duration in ms based on 60FPS refresh rate).
+ */
+class CameraAnimation {
+  /**
+   * @param {number} durationMs
+   * @param {!Array<!CameraOrientation>} orientations
+   */
+  constructor(durationMs, orientations) {
+    this.maxFrame = 60 / 1000 * durationMs;
+    this.orientations = orientations;
+    this.currentHeadingIndex = 0;
+    this.currentFrame = 0;
+  }
+
+  /**
+   * @param {number} x
+   * @return {number}
+   * @private
+   */
+  easeInOutQuad_(x) {
+    return x < 0.5 ? 2 * x * x : 1 - Math.pow(-2 * x + 2, 2) / 2;
+  }
+
+  /**
+   * Returns the CameraOrientation for the next frame, or null if the animation
+   * is over.
+   * @return {?CameraOrientation}
+   */
+  getNextOrientation() {
+    if (this.currentHeadingIndex < 0 ||
+        this.currentFrame == this.maxFrame - 1) {
+      // Animation ended.
+      return null;
+    }
+    this.currentFrame++;
+    const framesPerSection = this.maxFrame / (this.orientations.length - 1);
```

Comment:
Does this need to be calculated on every frame?

Reply:
indeed. good catch